templates/cpp-template-default/CMakeLists.txt

```diff
@@ -83,100 +83,157 @@ else()
 endif()
 
 target_link_libraries(${APP_NAME} cocos2d)
 
-set(APP_BIN_DIR "${CMAKE_BINARY_DIR}/bin")
 
-set_target_properties(${APP_NAME} PROPERTIES
-     RUNTIME_OUTPUT_DIRECTORY  "${APP_BIN_DIR}")
+if(MSVC)
 
-if ( WIN32 )
-  #also copying dlls to binary directory for the executable to run
-  pre_build(${APP_NAME}
-    COMMAND ${CMAKE_COMMAND} -E remove_directory ${APP_BIN_DIR}/Resources
-    COMMAND ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/Resources ${APP_BIN_DIR}/Resources
-    COMMAND ${CMAKE_COMMAND} -E copy ${COCOS2D_ROOT}/external/win32-specific/gles/prebuilt/glew32.dll ${APP_BIN_DIR}/${CMAKE_BUILD_TYPE}
-	COMMAND ${CMAKE_COMMAND} -E copy ${COCOS2D_ROOT}/external/win32-specific/zlib/prebuilt/zlib1.dll ${APP_BIN_DIR}/${CMAKE_BUILD_TYPE}
+	set(APP_BIN_DIR "${CMAKE_BINARY_DIR}/bin/${PLATFORM_FOLDER}/$<CONFIG>")
+
+    #get our resources
+    add_custom_command(TARGET ${APP_NAME} PRE_BUILD
+            COMMAND ${CMAKE_COMMAND} -E copy_directory
+            ${CMAKE_CURRENT_SOURCE_DIR}/Resources ${APP_BIN_DIR})
+	    
+    # create a list of dlls to copy
+	file(GLOB THIRD_PARTY_DLLS
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/gles/prebuilt/glew32.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/zlib/prebuilt/zlib1.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/OggDecoder/prebuilt/libogg.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/OggDecoder/prebuilt/libvorbis.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/OggDecoder/prebuilt/libvorbisfile.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/MP3Decoder/prebuilt/libmpg123.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/OpenalSoft/prebuilt/OpenAL32.dll"
+		"${COCOS2D_ROOT}/external/${PLATFORM_FOLDER}-specific/icon/prebuilt/iconv.dll"
+		"${COCOS2D_ROOT}/external/tiff/prebuilt/${PLATFORM_FOLDER}/libtiff.dll"
 	)
-elseif( ANDROID )
+	
+	list(SORT THIRD_PARTY_DLLS)
+
+	# do the copying
+	foreach( file_i ${THIRD_PARTY_DLLS})
+		add_custom_command(TARGET ${APP_NAME} PRE_BUILD
+			COMMAND ${CMAKE_COMMAND} -E copy
+			${file_i} 
+			${APP_BIN_DIR}
+		)
+	endforeach( file_i )
+
+    #Visual Studio Defaults to wrong type
+    set_target_properties(${APP_NAME} PROPERTIES LINK_FLAGS "/SUBSYSTEM:WINDOWS"
+									RUNTIME_OUTPUT_DIRECTORY  "${APP_BIN_DIR}")
+
+elseif(MACOSX OR APPLE)
+    set_target_properties(${APP_NAME} PROPERTIES
+            MACOSX_BUNDLE 1
+            LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib/$<CONFIG>"
+            ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib/$<CONFIG>"
+            RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/bin/$<CONFIG>"
+            )
 
 else()
-  pre_build(${APP_NAME}
-    COMMAND ${CMAKE_COMMAND} -E remove_directory ${APP_BIN_DIR}/Resources
-    COMMAND ${CMAKE_COMMAND} -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/Resources ${APP_BIN_DIR}/Resources
-    )
+
+    set(APP_BIN_DIR "${CMAKE_BINARY_DIR}/bin/$<CONFIG>")
```

Comment:
what's `$(CONFIG)`? now the bin will be in `bin/debug/linux/Debug/MyGame` instead of `bin/debug/linux/MyGame`,  it breaks cocos command.

Reply:
i correct it in https://github.com/cocos2d/cocos2d-x/pull/18258, i think `bin/debug/linux/MyGame` is more reasonable.

Reply:
If you build by
```
cmake ..
make
```
bin dir will be just `bin/`. So all targets (debug, release etc) will be in same folder.

Reply:
it is strange, i think cocos command also use cmake to build and use make to run. May be cocos command copy the output to needed path then run it. If so, then i have to modify cocos command. But this will not be included in v3.16.

Reply:
yep, cocos command has a deploy stage, it will copy bin and resources generated by cmake into `project_root/bin/debug|release/linux` folder. If i change the codes in cocos, then we should modify all templates and all test cases. I will leave it in next version.

Reply:
In my CMakeLists.txt win32 and macos targets also use separate path fo debug|release. cocos console doesn't use CMake for win32 and macos, but i think its good idea use same path for all targets in CMakeLists.txt. So maybe you will change APP_BIN_DIR in those targets too?

Reply:
Also i think good idea change project name on `set(APP_NAME MyGame)` after execution `cocos new`command (MyGame -> %project_name%).
I would have done it myself on this PR, but I did not understand very well cocos console logic.

Reply:
The copy logic of linux is [here](https://github.com/cocos2d/cocos2d-console/blob/v3/plugins/plugin_compile/project_compile.py#L1468), we can just modify the `result_dir `. Other platforms will have similar logic. You can try it yourself. I don't use cmake for other platforms except linux. But i think it is good if we can use cmake to generate VS|Xcode projects dynamically. That's what i want to have, but i don't have time to do it.

The `cocos new` command logic is [here](https://github.com/cocos2d/cocos2d-console/blob/v3/plugins/plugin_new/project_new.py).